### PR TITLE
[MoonEP] Use DeepGEMM psum layout, zero-copy shard and other optimization

### PR DESCRIPTION
**Pull request by lyppg**
This PR adds several optimization to improve MoonEP+DeepGEMM performance for Kimi K3:
1. Use DeepGEMM psum layout
2. Zero-copy dispatch/combine
3. Padding tokens no longer route to expert 0

## Result

Steady-state decode, greedy, 128 output tokens, `decode_bench.py`:

| concurrency | before (tok/s) | after (tok/s) | after + DSPARK γ=7 (tok/s) |
|---|---|---|---|
| 1 (per request) | 6.06 | 15.9 (2.6x) | 37.6 (6.2x) |
| 8 (aggregate) | 48.5 | 128.5 (2.65x) | 239 (4.9x) |
| 32 (aggregate) | 190 | 458 (2.4x) | 663 (3.5x) |

Decode step at c=8: 165 ms -> 62 ms. One MoE layer: 1.69 ms -> 0.62 ms.
gsm8k (200 questions, greedy): 0.995 before, 0.990 after, 0.995 with DSPARK.

---
### CI States

Latest PR Test (Base): :x: [Run #34829990237](https://github.com/sgl-project/sglang/actions/runs/34829990237)
Latest PR Test (Extra): :x: [Run #34829989726](https://github.com/sgl-project/sglang/actions/runs/34829989726)
Latest PR Test (AMD ROCm 10): :x: [Run #34829989732](https://github.com/sgl-project/sglang/actions/runs/34829989732)

### CODE REVIEW

File: python/sglang/srt/layers/moe/token_dispatcher/moonep.py [hunk 1]
```diff
@@ -612,8 +620,18 @@ def _pad_to_capacity(
             return hidden_states, topk_ids, topk_weights, num_tokens
 
         pad_tokens = capacity - num_tokens
+        top_k = topk_ids.shape[1]
         hidden_pad = hidden_states.new_zeros(pad_tokens, hidden_states.shape[1])
-        id_pad = topk_ids.new_zeros(pad_tokens, topk_ids.shape[1])
+        assert self.num_experts is not None
+        id_pad = (
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/token_dispatcher/moonep.py]: **[bug]** Dummy pad tokens are now given expert ids `arange(num_tokens * top_k, capacity * top_k) % num_experts` with zero weights. MoonEP treats those ids as real for `tokens_per_expert` / planning: every non-empty VM group is padded up to `token_padding` (default 128), and the buffer is only sized for about `2 * (E / EP)` non-empty segments (`NvS = S*K + (token_padding-1)*2*epn`). Spreading pads across the full expert space can mark dozens of otherwise-empty experts as occupied. Example: 1 real token, rung capacity 8, `top_k=8` yields 56 dummy assignments; with 384 experts / EP 64, `NvS` is ~1.5k rows while 56 groups * 128 already exceeds that. Planning then writes `cu_seqlens` past `NvS` (MoonEP’s own reference asserts this) and dispatch/combine read/write off the shard. Routing every pad to expert 0 was unbalanced; routing them through *all* experts overshoots the padding budget the other way.

**Suggestion:** Keep dummy assignments inside the existing non-empty set (repeat real `topk_ids` with zero weights), or park them on a small per-rank set that cannot exceed the `2*epn` non-empty-group bound. Do not `% num_experts` over the full table. Add a unit case: `num_tokens < capacity` must not increase the number of non-empty groups by more than the pad count’s collision with already-routed experts, and `cu_seqlens[-1] <= NvS`.

File: python/sglang/srt/layers/moe/token_dispatcher/moonep.py [hunk 2]
```diff
@@ -551,21 +562,18 @@ def __init__(
...
+        self._active_capacity = self.num_max_dispatch_tokens_per_rank
         self.num_prefetch_slots = None
 
     @staticmethod
     def _raise_unimplemented() -> NoReturn:
         raise NotImplementedError(_MOONEP_UNSUPPORTED_MESSAGE)
 
-    def _phase_capacity(self) -> int:
-        if self.decode_max_dispatch_tokens_per_rank is None:
+    def _phase_capacity(self, num_tokens: int) -> int:
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/token_dispatcher/moonep.py:572]: **[suggestion]** Decode capacity is now a power-of-two rung of the live token count (`8, 16, …, decode_cap`) instead of a single decode vs prefill choice via `get_is_extend_in_batch()`. `MoonEPBuffer` keys on `num_max_dispatch_tokens_per_rank`, so each rung allocates and keeps a full NVLink/VMM buffer. The first request that steps onto a new rung pays that alloc on the hot path, and CUDA-graph capture (`SGLANG_ENABLE_MOONEP_CUDA_GRAPH`) cannot replay a step whose capacity (hence buffer identity and `S`) differs from capture. Mixed prefill/decode batches whose token count is still `<= decode_cap` also silently take the decode-sized buffer.

**Suggestion:** If graphs or static shapes matter, pin decode to one captured `S` (the old decode_cap) and only vary padding. If rungs stay, destroy unused buffers or pre-create the needed keys, and document that graphs must capture the max rung.

File: python/sglang/srt/layers/moe/moe_runner/deep_gemm.py [hunk 3]
```diff
@@ -1334,44 +1618,60 @@ def pre_permute_moonep_to_deep_gemm(
...
+            dispatch_output.cu_seqlens,
+            dispatch_output.plan,
+            runner_config.num_experts,
+        )
+        psum, rows_end = _moonep_psum_layout(seg_start, seg_len)
+        running_state["moonep_segments"] = (
+            seg_start,
+            seg_len,
+            dispatch_output.hidden_states,
+            int(dispatch_output.capacity) * runner_config.top_k,
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/moe_runner/deep_gemm.py]: **[suggestion]** Finalize grid height is `capacity * top_k` (MoonEP `NvS_capacity`). That matches the planner’s per-destination real-token cap (`CAP = S*K`) only if the constructive balancer always drains surplus. A cheaper, planner-independent bound already in this function is `all_tokens` (`hidden_states.shape[0]`, i.e. `NvS`). Under-launching `_moonep_finalize_shard_kernel` would leave GEMM rows uncopied into the shard; combine would then send leftover dispatch activations.

**Suggestion:** Pass `all_tokens` (or `int(seg_len.max())` if a sync is acceptable) as `max_rows_per_group` so the grid is a property of the shard, not of planner folklore.

File: python/sglang/srt/layers/moe/moe_runner/deep_gemm.py [hunk 4]
```diff
@@ -1295,6 +1316,269 @@ def _moonep_finalize_rows(
...
+        r = row.to(tl.int64)
+        gate = tl.load(g_ptr + r * 2 * N + offs, mask=mask, other=0.0).to(tl.float32)
+        up = tl.load(g_ptr + r * 2 * N + N + offs, mask=mask, other=0.0).to(tl.float32)
+        # tanh(x) == 2*sigmoid(2x) - 1 (avoids a libdevice dependency)
+        gate_t = 2.0 * tl.sigmoid(2.0 * gate / situ_beta) - 1.0
+        gate = situ_beta * gate_t * tl.sigmoid(gate)
+        up_t = 2.0 * tl.sigmoid(2.0 * up / situ_linear_beta) - 1.0
+        y = gate * situ_linear_beta * up_t
+        amax = tl.clamp(tl.max(tl.abs(y), axis=1), min=1e-10, max=float("inf"))
+        # Same quantization as _situ_mul_quant_contig_kernel followed by
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/moe_runner/deep_gemm.py]: **[suggestion]** The three-line comment restates the next two stores (scale by `448/amax`, pack round-up exponent of `amax/448`) and names the old kernels. The code already shows the convention; the cross-kernel names go stale when those helpers move.

**Suggestion:** Drop it, or keep one clause that is not visible here: this fused kernel must stay bit-compatible with `_situ_mul_quant_contig_kernel` + `_cast_to_e8m0_with_rounding_up`, unlike `_moonep_quant_rows_kernel` which divides by the power-of-two scale.

File: python/sglang/srt/layers/moe/token_dispatcher/moonep.py [hunk 5]
```diff
@@ -690,14 +709,17 @@ def dispatch(
             topk_ids,
             tokens_per_expert,
             async_finish=False,
+            zero_copy=True,
         )
+        hidden_nvsh = hidden_nvsh.view(hidden_nvsh.shape)
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/token_dispatcher/moonep.py:734]: **[nit]** `hidden_nvsh = hidden_nvsh.view(hidden_nvsh.shape)` is a same-shape view: same storage, same `data_ptr()`, no contiguity change. Combine’s zero-copy check already uses `data_ptr()`, so this does not enable or disable zero-copy.

**Suggestion:** Delete it, or if a new Tensor object is required, comment the constraint (e.g. drop a MoonEP wrapper identity) in one line.

File: python/sglang/srt/layers/moe/token_dispatcher/moonep.py [hunk 6]
```diff
@@ -719,11 +741,14 @@ def combine(
                 f"MoonEPDispatcher.combine expected MOONEP input, got "
                 f"{combine_input.format}"
             )
-        hidden_states, _route_weights_sk, _event = self._get_buffer().combine(
+        buffer = self._get_buffer()
+        shard = buffer._require_ctx()["hidden_buf_local"]
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/token_dispatcher/moonep.py]: **[nit]** Combine probes the shard via `buffer._require_ctx()["hidden_buf_local"]`. Upstream MoonEP already exposes this as `hidden_nvsh_buffer_view` with the same aliasing rules.

**Suggestion:** Use the public property so a ctx-key rename does not break SGLang.

File: python/sglang/srt/layers/deep_gemm_wrapper/entrypoint.py [hunk 7]
```diff
@@ -131,13 +131,34 @@ def grouped_gemm_nt_bf16_masked(
...
+    if _psum_layout_checked:
+        return
+    import inspect
+
+    params = inspect.signature(deep_gemm.m_grouped_fp8_gemm_nt_contiguous).parameters
+    if "use_psum_layout" not in params:
+        raise RuntimeError(
+            "This MoE path hands DeepGEMM a prefix-sum grouped layout "
+            "(m_grouped_fp8_gemm_nt_contiguous(use_psum_layout=...)), which the "
+            "installed DeepGEMM does not accept. Install DeepGEMM >= 0.1.5."
```
Comment (ch-wan) [inline python/sglang/srt/layers/deep_gemm_wrapper/entrypoint.py]: **[nit]** The runtime check is “does `m_grouped_fp8_gemm_nt_contiguous` accept `use_psum_layout`?”, which is right, but the message says “Install DeepGEMM >= 0.1.5”. DeepGEMM 0.1.4.post1 already has that argument (`ensure_zero_padding` / `expected_m_for_psum_layout` too). Operators on the currently pinned wheel will get a version that cannot be satisfied by the number in the error.

**Suggestion:** Tell them to install a DeepGEMM build whose contiguous grouped GEMM accepts `use_psum_layout`, without a version pin, or name the first release that actually lacks it.

**Review by ch-wan — COMMENTED** [review]
## Summary

The PR switches MoonEP+DeepGEMM to a prefix-sum grouped layout over the local expert block (home experts + prefetch slots), quantizes/activates only the real rows, and writes GEMM output back into the NVL shard so dispatch/combine can skip boundary copies. The psum pairing with `block_view` and MoonEP’s 128-token group padding is internally consistent, and the zero-copy combine `data_ptr` check matches MoonEP’s contract. The dominant correctness risk is the new pad-token expert IDs: cycling dummy tokens across *all* experts can create far more non-empty 128-padded VM groups than MoonEP sized `NvS` for, which can overflow the comm buffer. Secondary risks are per-rung NVLink buffer allocation (CUDA graphs / first-step stalls) and a few unexplained nits.

## Issue counts by severity

- bugs: 1
- suggestions: 3
- nits: 3

File: python/sglang/srt/layers/moe/moe_runner/deep_gemm.py [hunk 8]
```diff
@@ -1295,6 +1316,259 @@ def _moonep_finalize_rows(
...
 
+def _moonep_local_segments(
+    pool, cu_seqlens: torch.Tensor, plan, num_global_experts: int
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    epn = pool.num_local_experts
+    home = pool.ep_rank * epn
+    device = cu_seqlens.device
+    idx = torch.cat(
+        [
+            torch.arange(home, home + epn, device=device),
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/moe_runner/deep_gemm.py]: **[bug]** `_moonep_local_segments` builds the DeepGEMM prefix-sum from only `[ep_rank * epn, ep_rank * epn + epn)` and the prefetch tail. DeepGEMM reads group 0 from row 0 and group g from `[align(psum[g-1], BLOCK_M), psum[g])`, and `psum[g] = seg_start[g] + seg_len[g]` equals that only when those entries are the non-empty groups in `cu_seqlens` order. They are not. Dispatch returns `cu_seqlens` of length `E+B` (see the same contract `_expert_ids_from_plan` already uses: one row per global expert, then B slots). Planning keeps a remote expert in its global row whenever it is not one of the top-B prefetch picks, and this rank can receive a whole remote home group (`epn` experts). `SGLANG_MOONEP_NUM_PREFETCH_SLOTS` defaults to -1, which resolves to `min(4, epn)`, so any model with more than 4 local experts leaves remote segments in the prefix. Those rows sit before the home experts, `seg_start[0] != 0`, and the first local expert is multiplied from row 0 with the wrong weights. The omitted tokens are never finalized (`_moonep_finalize_into_shard` walks the same list), so combine ships the dispatched activations for them. `block_view` cannot cover those experts either: it only has `epn + B` rows, which is what replaced the full-pool `group_rows` index. The new segment test never puts tokens outside the home range, so it stays green.

**Suggestion:** Build the prefix sum over every non-empty `cu_seqlens` group in order (empty groups are already length 0), and index weights the way `group_rows` did, including remote pool rows. If the GEMM is intentionally local-plus-slots only, require `B >= epn` and fail the launch when any other group has `cu_seqlens[g] > cu_seqlens[g-1]`.

File: python/sglang/srt/layers/moe/moe_runner/deep_gemm.py [hunk 9]
```diff
@@ -1295,6 +1316,259 @@ def _moonep_finalize_rows(
...
+        starts.to(torch.int32).contiguous(),
+        (ends - starts - n_pad).to(torch.int32).contiguous(),
+    )
+
+
+def _moonep_psum_layout(
+    seg_start: torch.Tensor, seg_len: torch.Tensor
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    psum = (seg_start + seg_len).to(torch.int32).contiguous()
+    rows_end = ((psum[-1:] + 127) // 128 * 128).to(torch.int32).contiguous()
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/moe_runner/deep_gemm.py]: **[suggestion]** Group padding is hardcoded to 128, but the two sides of that contract are runtime values. MoonEP pads each non-empty segment with `SGLANG_MOONEP_TOKEN_PADDING` (default 128), and DeepGEMM's M-grouped `BLOCK_M` is `get_mk_alignment_for_contiguous_layout()` (also default 128, and 256 on SM100 once something calls the setter). `psum[g] = seg_start[g] + seg_len[g]` lines up with `align(previous_end, BLOCK_M)` only when those two are equal. A smaller token padding makes the next `psum` fall below `align(previous, BLOCK_M)`; the scheduler does that subtraction in `uint32` and the tile count wraps. A larger one leaves a gap, so the next expert is read from the middle of the previous segment. `rows_end` has the same literal, so the quant kernel also stops short of the last tile.

**Suggestion:** Use one alignment for the segment pad, `rows_end`, and the GEMM, and refuse to launch when `token_padding` and `get_mk_alignment_for_contiguous_layout()` differ.
